packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -76,77 +76,148 @@ contract MixinExchangeCore is
     /*
     * Core exchange functions
     */
-
-    /// @dev Fills the input order.
-    /// @param order Order struct containing order specifications.
-    /// @param takerTokenFillAmount Desired amount of takerToken to fill.
-    /// @param signature Proof of signing order by maker.
-    /// @return Total amount of takerToken filled in trade.
-    function fillOrder(
-        Order order,
-        uint256 takerTokenFillAmount,
+    
+    function orderStatus(
+        Order memory order,
         bytes signature)
         public
-        returns (uint256 takerTokenFilledAmount)
+        view
+        returns (
+            uint8 status,
+            bytes32 orderHash,
+            uint256 filledAmount)
     {
-        // Compute the order hash
-        bytes32 orderHash = getOrderHash(order);
-
-        // Check if order has been cancelled
-        if (cancelled[orderHash]) {
-            LogError(uint8(Errors.ORDER_CANCELLED), orderHash);
-            return 0;
-        }
-
-        // Validate order and maker only if first time seen
-        // TODO: Read filled and cancelled only once
-        if (filled[orderHash] == 0) {
-            require(order.makerTokenAmount > 0);
-            require(order.takerTokenAmount > 0);
-            require(isValidSignature(orderHash, order.makerAddress, signature));
-        }
-
-        // Validate taker
-        if (order.takerAddress != address(0)) {
-            require(order.takerAddress == msg.sender);
+        // Compute the order hash and fetch filled amount
+        orderHash = getOrderHash(order);
+        filledAmount = filled[orderHash];
+        
+        // Validations on order only if first time seen
+        if (filledAmount == 0) {
+            
+            // If order.takerTokenAmount is zero, then the order will always
+            // be considered filled because:
+            //    0 == takerTokenAmount == filledAmount
+            // Instead of distinguishing between unfilled and filled zero taker
+            // amount orders, we choose not to support them.
+            if(order.takerTokenAmount == 0) {
+                status = uint8(Errors.ORDER_INVALID);
+                return;
```

Comment:
can we explicitly `return status` in all of these if statements?

Reply:
Don't think we can since we're returning `status`, `orderHash` and `filledAmount`.

Reply:
Oops, we can still `return (status, orderHash, 0)`